src/vcpkg/registries.cpp

```diff
@@ -1147,15 +1147,75 @@ namespace vcpkg
 
     const RegistryImplementation* RegistrySet::registry_for_port(StringView name) const
     {
-        for (const auto& registry : registries())
+        auto candidates = registries_for_port(name);
+        auto registry = candidates.front();
+        return registry.implementation;
+    }
+
+    int compare_package_prefix(StringView name, StringView prefix)
+    {
+        if (prefix.empty() || name.empty()) return -1;
+
+        size_t idx = 0;
+        auto cur = prefix.begin();
+        auto last = prefix.end();
+        for (; cur != last; ++idx, ++cur)
+        {
+            const auto c = *cur;
+            if (c == '*')
+            {
+                return (++cur == last) ? static_cast<int>(idx) : -1;
+            }
+
+            if (idx >= name.size()) return -1;
+            if (c != name[idx]) return -1;
+        }
+
+        // exact match, increase priority by 1
+        if (idx == name.size()) return name.size() + 1;
+        return static_cast<int>(idx);
```

Comment:
How about:

```diff
diff --git a/include/vcpkg/registries.h b/include/vcpkg/registries.h
index 799648a0..e635b1b7 100644
--- a/include/vcpkg/registries.h
+++ b/include/vcpkg/registries.h
@@ -103,13 +103,6 @@ namespace vcpkg
         std::unique_ptr<RegistryImplementation> implementation_;
     };
 
-    struct RegistryCandidate
-    {
-        const RegistryImplementation* implementation;
-        int priority;
-        size_t index;
-    };
-
     // this type implements the registry fall back logic from the registries RFC:
     // A port name maps to one of the non-default registries if that registry declares
     // that it is the registry for that port name, else it maps to the default registry
@@ -132,7 +125,7 @@ namespace vcpkg
         //
         // It is guaranteed to at least contain one element, the default registry.
         // NOTE: the default registry implementation can be null.
-        std::vector<RegistryCandidate> registries_for_port(StringView name) const;
+        std::vector<const RegistryImplementation*> registries_for_port(StringView name) const;
 
         ExpectedL<Version> baseline_for_port(StringView port_name) const;
 
@@ -212,6 +205,5 @@ namespace vcpkg
         VersionDbEntryDeserializer underlying;
     };
 
-    // exposed for testing
-    int compare_package_prefix(StringView name, StringView pattern);
+    size_t package_match_prefix(StringView name, StringView pattern);
 }
diff --git a/src/vcpkg-test/registries.cpp b/src/vcpkg-test/registries.cpp
index ec0c751b..0c1118f7 100644
--- a/src/vcpkg-test/registries.cpp
+++ b/src/vcpkg-test/registries.cpp
@@ -118,19 +118,19 @@ TEST_CASE ("check valid package patterns", "[registries]")
 
 TEST_CASE ("calculate prefix priority", "[registries]")
 {
-    CHECK(compare_package_prefix("boost", "*") == 0);
-    CHECK(compare_package_prefix("boost", "b*") == 1);
-    CHECK(compare_package_prefix("boost", "boost*") == 5);
-    CHECK(compare_package_prefix("boost", "boost") == 6);
-
-    CHECK(compare_package_prefix("", "") == -1);
-    CHECK(compare_package_prefix("", "*") == -1);
-    CHECK(compare_package_prefix("", "a") == -1);
-    CHECK(compare_package_prefix("boost", "") == -1);
-    CHECK(compare_package_prefix("boost", "c*") == -1);
-    CHECK(compare_package_prefix("boost", "*c") == -1);
-    CHECK(compare_package_prefix("boost", "c**") == -1);
-    CHECK(compare_package_prefix("boost", "c*a") == -1);
+    CHECK(package_match_prefix("boost", "*") == 1);
+    CHECK(package_match_prefix("boost", "b*") == 2);
+    CHECK(package_match_prefix("boost", "boost*") == 6);
+    CHECK(package_match_prefix("boost", "boost") == SIZE_MAX);
+
+    CHECK(package_match_prefix("", "") == SIZE_MAX);
+    CHECK(package_match_prefix("", "*") == 1);
+    CHECK(package_match_prefix("", "a") == 0);
+    CHECK(package_match_prefix("boost", "") == 0);
+    CHECK(package_match_prefix("boost", "c*") == 0);
+    CHECK(package_match_prefix("boost", "*c") == 0);
+    CHECK(package_match_prefix("boost", "c**") == 0);
+    CHECK(package_match_prefix("boost", "c*a") == 0);
 }
 
 TEST_CASE ("select highest priority registry", "[registries]")
@@ -183,19 +183,19 @@ TEST_CASE ("sort candidate registries by priority", "[registries]")
         REQUIRE(candidates.size() == 4);
         size_t idx = 0;
 
-        auto reg = candidates[idx++].implementation;
+        auto reg = candidates[idx++];
         REQUIRE(reg);
         CHECK(get_tri_num(*reg) == 4);
 
-        reg = candidates[idx++].implementation;
+        reg = candidates[idx++];
         REQUIRE(reg);
         CHECK(get_tri_num(*reg) == 3);
 
-        reg = candidates[idx++].implementation;
+        reg = candidates[idx++];
         REQUIRE(reg);
         CHECK(get_tri_num(*reg) == 1);
 
-        reg = candidates[idx++].implementation;
+        reg = candidates[idx++];
         REQUIRE(reg);
         CHECK(get_tri_num(*reg) == 2);
     }
@@ -209,9 +209,7 @@ TEST_CASE ("sort candidate registries by priority", "[registries]")
         RegistrySet set(nullptr, std::move(rs));
 
         auto candidates = set.registries_for_port("cpprestsdk");
-        REQUIRE(candidates.size() == 1);
-        auto reg = candidates.front().implementation;
-        CHECK_FALSE(reg);
+        REQUIRE(candidates.empty());
     }
 }
 
diff --git a/src/vcpkg/registries.cpp b/src/vcpkg/registries.cpp
index 820aeb42..2dcb26df 100644
--- a/src/vcpkg/registries.cpp
+++ b/src/vcpkg/registries.cpp
@@ -1148,74 +1148,71 @@ namespace vcpkg
     const RegistryImplementation* RegistrySet::registry_for_port(StringView name) const
     {
         auto candidates = registries_for_port(name);
-        auto registry = candidates.front();
-        return registry.implementation;
+        if (candidates.empty())
+        {
+            return default_registry();
+        }
+
+        return candidates[0];
     }
 
-    int compare_package_prefix(StringView name, StringView prefix)
+    size_t package_match_prefix(StringView name, StringView prefix)
     {
-        if (prefix.empty() || name.empty()) return -1;
+        if (name == prefix)
+        {
+            // exact match is like matching "infinity" prefix
+            return SIZE_MAX;
+        }
 
-        size_t idx = 0;
-        auto cur = prefix.begin();
-        auto last = prefix.end();
-        for (; cur != last; ++idx, ++cur)
+        // Note that the * is included in the match so that 0 means no match
+        const auto prefix_size = prefix.size();
+        if (prefix_size != 0)
         {
-            const auto c = *cur;
-            if (c == '*')
+            const auto star_index = prefix_size - 1;
+            if (prefix[star_index] == '*' && name.size() >= star_index &&
+                name.substr(0, star_index) == prefix.substr(0, star_index))
             {
-                return (++cur == last) ? static_cast<int>(idx) : -1;
+                return prefix_size;
             }
-
-            if (idx >= name.size()) return -1;
-            if (c != name[idx]) return -1;
         }
 
-        // exact match, increase priority by 1
-        if (idx == name.size()) return name.size() + 1;
-        return static_cast<int>(idx);
+        return 0;
     }
 
-    std::vector<RegistryCandidate> RegistrySet::registries_for_port(StringView name) const
+    std::vector<const RegistryImplementation*> RegistrySet::registries_for_port(StringView name) const
     {
-        size_t registry_index = 0;
+        struct RegistryCandidate
+        {
+            const RegistryImplementation* impl;
+            std::size_t matched_prefix;
+        };
+
         std::vector<RegistryCandidate> candidates;
-        for (auto it = registries().begin(); it != registries().end(); ++it, ++registry_index)
+        for (auto&& registry : registries())
         {
-            int priority = -1;
-            for (auto&& pattern : it->packages())
+            std::size_t longest_prefix = 0;
+            for (auto&& package : registry.packages())
             {
-                priority = std::max(priority, compare_package_prefix(name, pattern));
+                longest_prefix = std::max(longest_prefix, package_match_prefix(name, package));
             }
 
-            if (priority >= 0)
+            if (longest_prefix != 0)
             {
-                candidates.emplace_back(RegistryCandidate{
-                    &it->implementation(),
-                    priority,
-                    registry_index,
-                });
+                candidates.push_back({&registry.implementation(), longest_prefix});
             }
         }
 
         if (candidates.empty())
         {
-            candidates.emplace_back(RegistryCandidate{
-                default_registry(),
-                0,
-                0,
-            });
+            return std::vector<const RegistryImplementation*>();
         }
-        else
-        {
-            Util::sort(candidates, [](auto&& lhs, auto&& rhs) {
-                if (lhs.priority > rhs.priority) return true;
-                if (lhs.priority == rhs.priority) return lhs.index < rhs.index;
-                return false;
+
+        std::stable_sort(
+            candidates.begin(), candidates.end(), [](const RegistryCandidate& lhs, const RegistryCandidate& rhs) {
+                return lhs.matched_prefix > rhs.matched_prefix;
             });
-        }
 
-        return candidates;
+        return Util::fmap(std::move(candidates), [](const RegistryCandidate& target) { return target.impl; });
     }
 
     ExpectedL<Version> RegistrySet::baseline_for_port(StringView port_name) const
```

Reply:
These changes seem fair, and we don't expose the prioritization details.